x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/CompositeAggregationDataExtractor.java

```diff
@@ -0,0 +1,177 @@
+        Client client,
+        CompositeAggregationDataExtractorContext dataExtractorContext,
+        DatafeedTimingStatsReporter timingStatsReporter,
+        AggregatedSearchRequestBuilder requestBuilder
+    ) {
+        this.compositeAggregationBuilder = compositeAggregationBuilder;
+        this.client = Objects.requireNonNull(client);
+        this.context = Objects.requireNonNull(dataExtractorContext);
+        this.timingStatsReporter = Objects.requireNonNull(timingStatsReporter);
+        this.requestBuilder = Objects.requireNonNull(requestBuilder);
+        this.hasNext = true;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return hasNext;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return isCancelled;
+    }
+
+    @Override
+    public void cancel() {
+        LOGGER.debug(() -> new ParameterizedMessage("[{}] Data extractor received cancel request", context.jobId));
+        isCancelled = true;
+        hasNext = false;
+    }
+
+    @Override
+    public long getEndTime() {
+        return context.end;
+    }
+
+    @Override
+    public Optional<InputStream> next() throws IOException {
+        if (hasNext() == false) {
+            throw new NoSuchElementException();
+        }
+
+        Aggregations aggs = search();
+        if (aggs == null) {
+            LOGGER.trace(() -> new ParameterizedMessage("[{}] extraction finished", context.jobId));
+            hasNext = false;
+            afterKey.set(null);
+            return Optional.empty();
+        }
+        return Optional.of(processAggs(aggs));
+    }
+
+    private Aggregations search() {
+        long histogramSearchStartTime = Math.max(
+            0,
+            context.start - ExtractorUtils.getHistogramIntervalMillis(compositeAggregationBuilder)
+        );
+        LOGGER.trace(
+            () -> new ParameterizedMessage(
+                "[{}] Executing composite aggregated search from [{}] to [{}]",
+                context.jobId,
+                histogramSearchStartTime,
+                context.end
+            )
+        );
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+            .size(0)
+            .query(ExtractorUtils.wrapInTimeRangeQuery(context.query, context.timeField, histogramSearchStartTime, context.end));
+
+        if (context.runtimeMappings.isEmpty() == false) {
+            searchSourceBuilder.runtimeMappings(context.runtimeMappings);
+        }
+        if (afterKey.get() != null) {
+            compositeAggregationBuilder.aggregateAfter(afterKey.get());
+        }
+        searchSourceBuilder.aggregation(compositeAggregationBuilder);
+        ActionRequestBuilder<SearchRequest, SearchResponse> searchRequest = requestBuilder.build(searchSourceBuilder);
+        SearchResponse searchResponse = executeSearchRequest(searchRequest);
+        LOGGER.trace(() -> new ParameterizedMessage("[{}] Search composite response was obtained", context.jobId));
```

Comment:
Could be helpful to include the response itself in this log message here

Reply:
@dimitris-athanasiou maybe, but the search response would have data from the documents. While we do this elsewhere, I am not 100% sure we should do this at all. 

Basically, if another cluster is monitoring the ES logs, we leak the data that is stored in this ES cluster if we log like that. Doesn't seem prudent